Added implementation for NSUbiquitousKeyValueStore

arthef opened the pull request:
Using `NSUbiquitousKeyValueStore` settings can be automatically shared between devices. The implantation follows convention of other implementations and is mainly based on the `NSUserDefaultsSettings` code.

Except standard initialization, no other special code is needed. However, to make key-value storage work, the app needs to be correctly configured in Xcode:

In the app project, TARGETS, Signing & Capabilities a new capability needs to be added (`+ Capability`):
1. iCloud, Service: __Key-value storage__

Also, it is highly recommended to add and implement external change listener(s) to make the app aware of the settings change on other device.

Commit:
Added support for NSUbiquitousKeyValueStore on iOS/MacOS

Commit:
Corrected code documentation and removed caching from change listener callbacks.

russhwolf commented on multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt:
```diff
@@ -0,0 +1,239 @@
+ * delegate, or via a [Factory].
+ */
+public class NSUbiquitousKeyValueStoreSettings public constructor(
+    private val delegate: NSUbiquitousKeyValueStore
```
This can use the default value `NSUbiquitousKeyValueStore.defaultStore` since that will almost always be what users want.

```suggestion
    private val delegate: NSUbiquitousKeyValueStore = NSUbiquitousKeyValueStore.defaultStore
```

russhwolf commented on multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt:
```diff
@@ -0,0 +1,239 @@
+public class NSUbiquitousKeyValueStoreSettings public constructor(
+    private val delegate: NSUbiquitousKeyValueStore
+) : ObservableSettings {
+
+    /**
+     * A factory that can produce [Settings] instances.
+     *
+     * This class creates `Settings` objects backed by [NSUbiquitousKeyValueStore].
+     */
+    public class Factory : Settings.Factory {
+
+        /**
+         * Creates a [Settings] object associated with the provided [name].
+         *
+         * Multiple `Settings` instances created with the same `name` parameter will be backed by the same persistent
+         * data, while distinct `name`s will use different data. If `name` is `null` then a platform-specific default
+         * will be used.
+         *
+         * On the iOS and macOS platforms, this is implemented by calling [NSUbiquitousKeyValueStore.defaultStore]
+         */
+        public override fun create(name: String?): NSUbiquitousKeyValueStoreSettings {
+            val delegate = NSUbiquitousKeyValueStore.defaultStore
+            return NSUbiquitousKeyValueStoreSettings(delegate)
+        }
+    }
```
The `Factory` interface is only for when it's possible to name different delegate instances. Since `NSUbiquitousKeyValueStore` only allows using the one `defaultStore` instance, the class should not include a `Factory`.

```suggestion
```

russhwolf commented on multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt:
```diff
@@ -0,0 +1,239 @@
+            callback.invoke()
+        }
+        val observer = NSNotificationCenter.defaultCenter.addObserverForName(
+            name = NSUbiquitousKeyValueStoreDidChangeExternallyNotification,
```
I'm not certain this will work as intended. The documentation for `NSUbiquitousKeyValueStoreDidChangeExternallyNotification` suggests that notifications are only served from external changes, meaning that if we update a value in-app it won't propagate to listeners. This would break the expectations of `ObservableSettings`

arthef replied:
You are correct here. The notification does not trigger to the app which made the change, which is, as I understand, your intended function.
The notification from the `NSUbiquitousKeyValueStore` changes is only triggered for apps running on other devices.
However, I could not see any other place or way or API to implement external changes notification for the app, therefore I used this one.
From what I read in the documentation, the `NSUbiquitousKeyValueStore` does not send notification to the local app at all.

If you have any idea or suggestion on how to implement this in a different way, please let me know. I can rewrite the code.

russhwolf replied:
I'll have to think about this some. I'm not sure whether it makes sense to use `ObservableSettings` instead of `Settings` here. Do we eventually see updated values on get calls even without subscribing to external notifications?

russhwolf commented on multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt:
```diff
@@ -0,0 +1,239 @@
+    }
+}
+
+internal expect fun NSUbiquitousKeyValueStore.intForKey(defaultName: String): Int
+internal expect fun NSUbiquitousKeyValueStore.setInt(value: Int, forKey: String)
+internal expect fun NSUbiquitousKeyValueStore.longForKey(defaultName: String): Long
+internal expect fun NSUbiquitousKeyValueStore.setLong(value: Long, forKey: String)
+internal expect fun NSUbiquitousKeyValueStore.setFloat(value: Float, forKey: String)
+internal expect fun NSUbiquitousKeyValueStore.floatForKey(defaultName: String): Float
```
There's no reason to use expect/actual for these if the implementations are the same for 32-bit and 64-bit systems.

arthef replied:
As I said, not fully understanding the library implementation I tried to keep it as close to the `NSUserDefaults` as possible. I can rewrite this part if you want. Please confirm.

russhwolf replied:
Yes, please rewrite it without expect/actual

arthef replied:
Done

russhwolf commented on multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt:
```diff
@@ -0,0 +1,239 @@
+
+    private fun addListener(key: String, callback: () -> Unit): SettingsListener {
+        val block = { _: NSNotification? ->
+            callback.invoke()
```
This probably should be caching the previous value, similar to other implementations, in order to avoid triggering the listener when nothing has changed.

arthef replied:
I ran several tests and have it working in my app. It looks like the `NSUbiquitousKeyValueStore` takes care of avoiding notifications if nothing changed. So, it basically handles it itself efficiently. Therefore, from my experience, caching here is not needed for the `NSUbiquitousKeyValueStore` implementation. However, I can put it back for consistency with other implementations.

Please confirm that this is what you want.

russhwolf replied:
The issue is, if you have two different keys "a" and "b" but you're only observing "a", the notification will trigger for changes to both "a" and "b". The cache ensures that the update callbacks only trigger when the key you're observing changes.

arthef replied:
I added cache back and confirmed it works as expected during my tests.

russhwolf reviewed (changes requested):
Thanks for the contribution! I've never worked with `NSUbiquitousKeyValueStore` so I want to make sure everything works as expected before merging this.

The most important item to that end is to make sure this class passes the test suite for `Settings` implementations, in order to verify that it behaves as expected. This means adding a test class that extends `BaseSettingsTest`. Because it requires entitlements, this takes a bit of setup. You can use the `keychain-tests` module as a template.

I have other priorities at the moment so I don't expect to get this merged before the next library release (1.4.0).

arthef committed:
Update multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt

Yes, I agree, this makes sense.

Co-authored-by: Russell Wolf <russhwolf@users.noreply.github.com>

arthef committed:
Update multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/NSUbiquitousKeyValueStoreSettings.kt

Ok, TBH, when I was working on the code, I did not fully understand the existing implementation and therefore, tried to keep the new class as close to the NSUserDefaults implementation as possible.

Co-authored-by: Russell Wolf <russhwolf@users.noreply.github.com>

Commit:
Added cache back to the update listeners. Confirmed it works.

Commit:
Rewritten code without unnecessary expect/actual as requested.

arthef commented:
Hopefully with my recent commits all concerns are resolved.
Please take a look and let me know if there is anything else to be done to accept the PR.

russhwolf commented:
I'd like to get tests running against this as well. I can take a pass at it at some point but probably not soon.

coreypett-cube commented:
Any ETA on this? Soon to fork and apply @arthef's work.

russhwolf commented:
No ETA, sorry. I don't want to merge this without having tests running against it. Open to contributions there.

If you want it sooner you should be able to just copy `NSUbiquitousKeyValueStoreSettings.kt` to your project without needing to fork anything.
